xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py

```diff
@@ -0,0 +1,550 @@
+            mutable_blocks.append(mutable_block := Block.from_arg_types(block.arg_types))
+            block_mapping[block] = mutable_block
+        for block in self.blocks:
+            # This will use the already created Block and populate it
+            block.get_mutable_copy(value_mapping=value_mapping, block_mapping=block_mapping)
+        return Region.from_block_list(mutable_blocks)
+
+
+@dataclass(frozen=True)
+class IBlock:
+    """An immutable block contains a list of immutable operations. IBlocks are contained in IRegions."""
+    args: IList[IBlockArg]
+    ops: IList[IOp]
+
+    @property
+    def arg_types(self) -> List[Attribute]:
+        frozen_arg_types = [arg.typ for arg in self.args]
+        return frozen_arg_types
+
+    def __hash__(self) -> int:
+        return (id(self))
+
+    def __eq__(self, __o: object) -> bool:
+        return self is __o
+
+    def __repr__(self) -> str:
+        return "block of" + str(len(self.ops)) + " with args: " + str(
+            self.args)
+
+    def __post_init__(self):
+        for arg in self.args:
+            object.__setattr__(arg, "block", self)
+
+    def __init__(self, args: Sequence[Attribute] | Sequence[IBlockArg],
+                 ops: Sequence[IOp]):
+        """Creates a new immutable block."""
+
+        # Type Guards:
+        def is_iblock_arg_seq(
+                list: Sequence[Any]) -> TypeGuard[Sequence[IBlockArg]]:
+            if len(list) == 0:
+                return False
+            return all([isinstance(elem, IBlockArg) for elem in list])
+
+        def is_type_seq(list: Sequence[Any]) -> TypeGuard[Sequence[Attribute]]:
+            return all([isinstance(elem, Attribute) for elem in list])
+
+        if is_type_seq(args):
+            block_args: Sequence[IBlockArg] = [
+                IBlockArg(type, IList([]), self, idx) for idx, type in enumerate(args)
+            ]
+        elif is_iblock_arg_seq(args):
+            block_args: Sequence[IBlockArg] = args
+            for block_arg in block_args:
+                object.__setattr__(block_arg, "block", self)
+        else:
+            raise Exception("args for IBlock ill structured")
+
+        object.__setattr__(self, "args", IList(block_args))
+        object.__setattr__(self, "ops", IList(ops))
+
+        self.args.freeze()
+        self.ops.freeze()
+
+    @classmethod
+    def from_mutable(
+        cls,
+        block: Block,
+        value_map: Optional[dict[SSAValue, ISSAValue]] = None,
+        block_map: Optional[dict[Block, IBlock]] = None,
+    ) -> IBlock:
+        """
+        Creates an immutable block from a mutable block. 
+        The value_map and block_map are used to map already known correspondings
+        of mutable values to immutable values and mutable blocks to immutable blocks.
+        """
+        if value_map is None:
+            value_map = {}
+        if block_map is None:
+            block_map = {}
+
+        args: List[IBlockArg] = []
```

Comment:
The way I see it, you are putting dummy `IBlockArg`, so users can point to it when creating the operations in `Block`.
Instead, could we create the `IBlock` with the arguments, and then add the operations?
That way, we don't need this "hack" of having `IBlockArg` with a `None` block.

Reply:
Yes, thank you. I restructured it